tests/basictests.cpp

```diff
@@ -24,479 +111,549 @@ inline uint64_t f64_ulp_dist(double a, double b) {
       uint64_t x = pjh.get_unsigned_integer();
       int power = 0;
-      while(x > 1) {
-         if((x % 10) != 0) {
-           printf("failed to parse %s. \n", buf);
-           return false;
-         }
-         x = x / 10;
-         power ++;
+      while (x > 1) {
+        if ((x % 10) != 0) {
+          printf("failed to parse %s. \n", buf);
+          return false;
+        }
+        x = x / 10;
+        power++;
       }
-      if(power != i) {
-         printf("failed to parse %s. \n", buf);
-         return false;
+      if (power != i) {
+        printf("failed to parse %s. \n", buf);
+        return false;
       }
     } else {
       double x = pjh.get_double();
       double expected = std::pow(10, i);
-      int ulp = (int) f64_ulp_dist(x, expected);
-      if(ulp > 1) {
-         printf("failed to parse %s. \n", buf);
-         printf("actual: %.20g expected: %.20g \n", x, expected);
-         printf("ULP: %d \n", ulp);
-         return false;
+      int ulp = (int)simdjson::lib::f64_ulp_dist(x, expected);
+      if (ulp > 1) {
+        printf("failed to parse %s. \n", buf);
+        printf("actual: %.20g expected: %.20g \n", x, expected);
+        printf("ULP: %d \n", ulp);
+        return false;
       }
     }
   }
   printf("Powers of 10 can be parsed.\n");
   return true;
 }
 
-
-// adversarial example that once triggred overruns, see https://github.com/lemire/simdjson/issues/345
+// -------------------------------------------------------------------------------------------------
+// adversarial example that once triggred overruns, see
+// https://github.com/lemire/simdjson/issues/345
 bool bad_example() {
-  std::string badjson = "[7,7,7,7,6,7,7,7,6,7,7,6,[7,7,7,7,6,7,7,7,6,7,7,6,7,7,7,7,7,7,6";
+
+  using namespace std;
+  using namespace simdjson::lib::literals; 
+
+  constexpr auto badjson =
+      "[7,7,7,7,6,7,7,7,6,7,7,6,[7,7,7,7,6,7,7,7,6,7,7,6,7,7,7,7,7,7,6"_sv;
+
   simdjson::ParsedJson pj = simdjson::build_parsed_json(badjson);
-  if(pj.is_valid()) {
-    printf("This json should not be valid %s.\n", badjson.c_str());
+  if (pj.is_valid()) {
+    printf("This json should not be valid %s.\n", badjson.data());
     return false;
   }
   return true;
 }
+// -------------------------------------------------------------------------------------------------
 // returns true if successful
 bool stable_test() {
-  std::string json = "{"
-        "\"Image\":{"
-            "\"Width\":800,"
-            "\"Height\":600,"
-            "\"Title\":\"View from 15th Floor\","
-            "\"Thumbnail\":{"
-            "\"Url\":\"http://www.example.com/image/481989943\","
-            "\"Height\":125,"
-            "\"Width\":100"
-            "},"
-            "\"Animated\":false,"
-            "\"IDs\":[116,943.3,234,38793]"
-          "}"
-      "}";
+  using namespace std;
+  using namespace simdjson::lib::literals;
+  // std::string_view
+  // note: it seems currently (2020 Q1) simdjson input must not have white
```

Comment:
This is incorrect. The simdjson parser is standard compliant. Allowable spaces are fine. Between elements you can use tabs, line endings (\n, \r), spaces....

Reply:
When json is stored on the "tape" is it with or without white spaces?

That comment was left for us, of course, it will be removed.

Reply:
The JSON is not stored on the tape. The tape is a materialization of the DOM. However, from the DOM, you would not be able to reproduce the original document in general (whether there are spaces in it or not).

However, the input is definitively allowed to contain spaces.

Reply:
json (string) -> stage_1 -> stage_2 -> "tape" (array) -- is that the workflow?

Reply:
@DBJDBJ That is correct. Stage 1 will happily skip over spaces.

Reply:
> The JSON is not stored on the tape. The tape is a materialization of the DOM.

I assume the "Tape" (and its Iterator) is what users use from their C++ code as the final product of `simdjson` work.

I just went through your key document, once more.  I know what are you talking about, And I know the "Fig 1" ... But you have never specifically mentioned this "materialization of the DOM"  idea in there. In fact, I could not find the word "DOM" in that document.

The diagrammatic representation of the "Tape" should be the third column in Fig. 1, I think.

![image](https://user-images.githubusercontent.com/145155/72569926-d57cff00-38bb-11ea-99d9-e0585a861b91.png)


"Tape" is your data structure storing the DOM.  It is logically the DOM, but physically an array.
You are not using any of C++ std lib containers but a "special" structure which is basically an array.

Reply:
You may very well be correct if you are saying that this is not fully explained/documented.